### Настройки

Начинаем с импорта библиотек

In [ ]:
import pandas as pd
import psycopg2

In [ ]:
# хороший тон при фиксации изменений в github не публиковать код с открытыми данными для входа
# для запуска кода поместите данные, которые указаны на странице курса в эту ячейку
# при размещении в github удалите эту ячейку
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

Создаем соединение с заданными параметрами

In [ ]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

In [ ]:
# код запроса представляет из себя строковую переменную
n = 10
query = f'''
            SELECT *
            FROM public.vacancies
            LIMIT {n}
         '''
# лучше использовать f-строки - так удобнее передавать параметры

### Выполнение запроса (проверка работоспособности)



In [ ]:
# используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df = pd.read_sql_query(query, connection)
df

/tmp/ipython-input-2473184231.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


id                                               name  \
0  55312386                                Компьютерный Мастер   
1  55843014                            Системный администратор   
2  54525964                      Lead Java Developer to Poland   
3  54525965                      Lead Java Developer to Poland   
4  55354053  Специалист службы поддержки с техническими зна...   
5  55906072                                          Корректор   
6  55523207                   React Front-end Middle Developer   
7  55050261                                Инженер-программист   
8  55583432                            Системный администратор   
9  55312414        Ведущий специалист группы разработки систем   

                                          key_skills          schedule  \
0  Пользователь ПК\tРабота в команде\tРемонт ноут...       Полный день   
1  Средства криптографической защиты информации\t...       Полный день   
2    Spring Framework\tSQL\tHibernate ORM\tJava\tGit  Удаленная работа   
3    Spring Framework\tSQL\tHibernate ORM\tJava\tGit  Удаленная работа   
4                                               None  Удаленная работа   
5  Корректура текстов\tГрамотность\tРабота в кома...       Полный день   
6       JavaScript\tTypeScript\tCSS3\tNode.js\tHTML5     Гибкий график   
7  Java SE\tSpring Framework\tSpring Boot\tGit\tS...       Полный день   
8  Linux\tАдминистрирование сетевого оборудования...       Полный день   
9  1С программирование\t1С: Предприятие 8\t1С: До...       Полный день   

           experience           employment  salary_from  salary_to  area_id  \
0           Нет опыта     Полная занятость      64000.0        NaN     1516   
1       От 3 до 6 лет     Полная занятость          NaN        NaN      130   
2       От 3 до 6 лет     Полная занятость          NaN        NaN      160   
3       От 3 до 6 лет     Полная занятость          NaN        NaN      159   
4           Нет опыта  Частичная занятость      15000.0        NaN     1955   
5           Нет опыта     Полная занятость          NaN        NaN     2323   
6  От 1 года до 3 лет     Полная занятость          NaN        NaN     2759   
7           Нет опыта     Полная занятость      40000.0    60000.0       49   
8  От 1 года до 3 лет     Полная занятость      50000.0        NaN        3   
9       От 3 до 6 лет     Полная занятость          NaN        NaN       47   

   employer_id  
0      5724811  
1      4903713  
2        69961  
3        69961  
4         1740  
5       584934  
6       204511  
7      4448636  
8        55126  
9      1413754

## Предварительный анализ
### Задание 1
Сколько вакансий в базе?

In [ ]:
query = f'''
            SELECT COUNT(*)
            FROM public.vacancies
         '''
df = pd.read_sql_query(query, connection)
df

/tmp/ipython-input-2666057266.py:4: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0  49197

### Задание 2

Сколько работодателей?

In [ ]:
query = f'''
            SELECT COUNT(*)
            FROM public.employers
         '''
df = pd.read_sql_query(query, connection)
df

/tmp/ipython-input-1419425662.py:4: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0  23501

### Задание 3
Сколько регионов?

In [ ]:
query = f'''
            SELECT COUNT(*)
            FROM public.areas
         '''
df = pd.read_sql_query(query, connection)
df

/tmp/ipython-input-2804089023.py:4: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0   1362

### Задание 3
Сколько сфер деятельности?

In [ ]:
query = f'''
            SELECT COUNT(*)
            FROM public.industries
         '''
df = pd.read_sql_query(query, connection)
df

/tmp/ipython-input-1997115098.py:4: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0    294

## Детальный анализ вакансий
### Задание 1
Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Выберите пятёрку лидеров по количеству вакансий:

In [ ]:
query = f'''
            SELECT
            a.name AS area,
            COUNT(v.id) AS cnt
            FROM public.vacancies v
            JOIN public.areas a ON v.area_id = a.id
            GROUP BY a.name
            ORDER BY cnt DESC
            LIMIT 5
         '''
df = pd.read_sql_query(query, connection)
df

/tmp/ipython-input-610969957.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


area   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892

### Задание 2
Посмотрим на зарплаты. У какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой?

In [ ]:
query = f'''
            SELECT COUNT(*)
            FROM public.vacancies
            WHERE salary_from IS NOT NULL OR salary_to IS NOT NULL;
         '''
df = pd.read_sql_query(query, connection)
df

/tmp/ipython-input-2660883360.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0  24073

### Задание 3
Найдите средние значения для нижней и верхней границы зарплатной вилки.
Округлите значения до целого числа.

In [ ]:
query = f'''
            SELECT ROUND(AVG(salary_from)) AS avg_salary_from,
                   ROUND(AVG(salary_to)) AS avg_salary_to
            FROM public.vacancies
            WHERE salary_from IS NOT NULL OR salary_to IS NOT NULL;
         '''
df = pd.read_sql_query(query, connection)
df

/tmp/ipython-input-3610087671.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


avg_salary_from  avg_salary_to
0          71065.0       110537.0

## Задание 4
Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Какая пара находится на втором месте по популярности?

In [ ]:
query = f'''
            SELECT schedule, employment, COUNT(*) AS cnt
            FROM public.vacancies
            GROUP BY schedule, employment
            ORDER BY cnt DESC;
         '''
df = pd.read_sql_query(query, connection)
df

/tmp/ipython-input-3109613216.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


schedule           employment    cnt
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

## Задание 5
Напишите запрос, выводящий значения поля «Требуемый опыт работы» (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта.
Выберите верный вариант ответа.

In [ ]:
query = f'''
            SELECT experience, COUNT(*) AS cnt
            FROM public.vacancies
            GROUP BY experience
            ORDER BY cnt ASC;
         '''
df = pd.read_sql_query(query, connection)
df

/tmp/ipython-input-981652000.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


experience    cnt
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

### Анализ работодателей
## Задание 1
Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [ ]:
query = f'''
            WITH ranked_employers AS (
                SELECT e.name AS employer_name, COUNT(v.id) AS vacancy_count,
                       ROW_NUMBER() OVER (ORDER BY COUNT(v.id) DESC) as rank
                FROM public.vacancies v
                JOIN public.employers e ON v.employer_id = e.id
                GROUP BY e.name
            )
            SELECT employer_name, vacancy_count
            FROM ranked_employers
            WHERE rank IN (1, 5)
            ORDER BY rank;
         '''
df = pd.read_sql_query(query, connection)
df

/tmp/ipython-input-2135745929.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


employer_name  vacancy_count
0         Яндекс           1933
1  Газпром нефть            331

## Задание 2
Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём. Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей. Впишите его название в поле ниже в том виде, который вернул запрос.

In [ ]:
query = f'''
            SELECT
                a.name AS area_name,
                COUNT(e.id) AS employer_count
            FROM public.areas a
            LEFT JOIN public.employers e ON a.id = e.area
            LEFT JOIN public.vacancies v ON a.id = v.area_id
            WHERE v.id IS NULL
            GROUP BY a.name
            ORDER BY COUNT(e.id) DESC
            LIMIT 1;
         '''
df = pd.read_sql_query(query, connection)
df

/tmp/ipython-input-3874297210.py:13: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


area_name  employer_count
0    Россия             410

### Задание 3
Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии.Выберите максимальное значение из получившегося списка.



In [ ]:
query = f'''
            SELECT MAX(region_count) AS max_regions
            FROM (
                SELECT e.name, COUNT(DISTINCT v.area_id) AS region_count
                FROM public.employers e
                JOIN public.vacancies v ON e.id = v.employer_id
                GROUP BY e.name
            ) subquery;
         '''
df = pd.read_sql_query(query, connection)
df

/tmp/ipython-input-260314907.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


max_regions
0          181

### Задание 4
Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. Введите количество, которое вернул запрос, в поле ниже.

In [ ]:
query = f'''
            SELECT COUNT(*)
            FROM public.employers e
            LEFT JOIN public.employers_industries ei ON e.id = ei.employer_id
            WHERE ei.industry_id IS NULL;
         '''
df = pd.read_sql_query(query, connection)
df

/tmp/ipython-input-2713174985.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0   8419

## Задание 5
Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. Введите в поле ниже название этой компании так же, как оно указано в результате запроса.

In [ ]:
query = f'''
            SELECT e.name
            FROM public.employers e
            JOIN public.employers_industries ei ON e.id = ei.employer_id
            GROUP BY e.id, e.name
            HAVING COUNT(ei.industry_id) = 4
            ORDER BY e.name
            LIMIT 1 OFFSET 2;
         '''
df = pd.read_sql_query(query, connection)
df

/tmp/ipython-input-102359749.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


name
0  2ГИС

## Задание 6
С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана «Разработка программного обеспечения».

In [ ]:
query = f'''
            SELECT COUNT(DISTINCT ei.employer_id)
            FROM public.employers_industries ei
            JOIN public.industries i ON ei.industry_id = i.id
            WHERE i.name = 'Разработка программного обеспечения';
         '''
df = pd.read_sql_query(query, connection)
df

/tmp/ipython-input-2297385087.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0   3553

## Задание 7

Для компании «Яндекс» выведите список городов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку "Total" с общим количеством вакансий компании, собранных в этой таблице. Должна получиться выборка такого вида (приведён пример результата для компании SberTech):

In [ ]:
query = f'''
             WITH yandex_vacancies_by_area AS (
    SELECT
        a.name AS city,
        COUNT(v.id) AS vacancy_count
    FROM public.vacancies v
    JOIN public.employers e ON v.employer_id = e.id
    JOIN public.areas a ON v.area_id = a.id
    WHERE e.name = 'Яндекс'
    GROUP BY a.name
),
million_cities AS (
    SELECT unnest(ARRAY[
        'Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань',
        'Нижний Новгород', 'Челябинск', 'Самара', 'Омск', 'Ростов-на-Дону',
        'Уфа', 'Красноярск', 'Пермь', 'Воронеж', 'Волгоград', 'Краснодар'
    ]) AS name
),
filtered_cities AS (
    SELECT
        y.city,
        y.vacancy_count
    FROM yandex_vacancies_by_area y
    INNER JOIN million_cities m ON y.city = m.name
),
total_row AS (
    SELECT 'Total' AS city, SUM(vacancy_count) AS vacancy_count
    FROM filtered_cities
)
SELECT city, vacancy_count
FROM (
    SELECT city, vacancy_count, 0 AS sort_order
    FROM filtered_cities
    UNION ALL
    SELECT city, vacancy_count, 1 AS sort_order
    FROM total_row
) combined
ORDER BY sort_order, city;
         '''

df = pd.read_sql_query(query, connection)
df

/tmp/ipython-input-296019539.py:41: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


city  vacancy_count
0         Волгоград           24.0
1           Воронеж           32.0
2      Екатеринбург           39.0
3            Казань           25.0
4        Красноярск           23.0
5            Москва           54.0
6   Нижний Новгород           36.0
7       Новосибирск           35.0
8              Омск           21.0
9             Пермь           25.0
10   Ростов-на-Дону           25.0
11           Самара           26.0
12  Санкт-Петербург           42.0
13              Уфа           26.0
14        Челябинск           22.0
15            Total          455.0

### Предметный анализ
## Задание 1
Сколько вакансий имеет отношение к данным?

In [ ]:
query = f'''
            SELECT COUNT(*)
FROM public.vacancies
WHERE (
    LOWER(name) LIKE '%data%'
    OR LOWER(name) LIKE '%данн%'
);
         '''
df = pd.read_sql_query(query, connection)
df

/tmp/ipython-input-1942247342.py:9: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0   1771

## Вывод
На рынке представленые 1771 вакансия тем или иным образом связанная с данными. Это говорит о том, что DS — устоявшаяся и востребованная профессия в России.

## Задание 2

Сколько есть подходящих вакансий для начинающего дата-сайентиста?
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
‘data scientist’;
‘data science’;
‘исследователь данных’;
‘ML’ (здесь не нужно брать вакансии по HTML);
‘machine learning’;
‘машинн%обучен%’.
В следующих заданиях мы продолжим работать с вакансиями по указанному выше условию (уже не учитывая вакансии уровня Junior).
Считаем вакансиями для специалистов уровня Junior следующие:
в названии есть слово “junior” или
требуемый опыт — «Нет опыта» или
тип трудоустройства — «Стажировка».
В качестве ответа запишите количество вакансий, которое вернул запрос.

In [ ]:
query = f'''
SELECT COUNT(*)
FROM public.vacancies
WHERE (
    LOWER(name) LIKE '%data scientist%'
    OR LOWER(name) LIKE '%data science%'
    OR LOWER(name) LIKE '%исследователь данных%'
    OR (LOWER(name) LIKE '%ml%' AND LOWER(name) NOT LIKE '%html%')
    OR LOWER(name) LIKE '%machine learning%'
    OR LOWER(name) LIKE '%машинн%обучен%'
)
AND (
    LOWER(name) LIKE '%junior%'
    OR experience = 'Нет опыта'
    OR employment = 'Стажировка'
);
'''
df = pd.read_sql_query(query, connection)
df

/tmp/ipython-input-1544103740.py:18: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0     51

## Вывод
Количество вакансия для начинающих специалистов в области данным крайне мало по сравнению с общим количеством вакансий. Это говорит о том, что
рынок насыщен опытными специалистами. Компании не готовы инвестировать в обучение новичков — они ищут "готовых к бою" кандидатов. Высокий порог входа в профессию. DS-рынок становится "зрелым", где новичкам всё сложнее пробиться без сильного портфолио или стажировки.

## Задание 3
Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или Postgres?
В качестве ответа запишите количество, которое вернул запрос.

In [ ]:
query = f'''
            SELECT COUNT(*)
            FROM public.vacancies
            WHERE (LOWER(name) LIKE '%data scientist%'
                   OR LOWER(name) LIKE '%data science%'
                   OR LOWER(name) LIKE '%исследователь данных%'
                   OR (LOWER(name) LIKE '%ml%' AND LOWER(name) NOT LIKE '%html%')
                   OR LOWER(name) LIKE '%machine learning%'
                   OR LOWER(name) LIKE '%машинн%обучен%')
              AND (LOWER(key_skills) LIKE '%sql%'
                   OR LOWER(key_skills) LIKE '%postgres%');
         '''
df = pd.read_sql_query(query, connection)
df

/tmp/ipython-input-18271579.py:13: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0    229

## Вывод

SQL нужный но необязательный навык, в основном от требуется в вакансиях связанных с аналитикой

## Задание 4
С помощью запроса, аналогичного предыдущему, проверьте, насколько популярен Python в требованиях работодателей к DS. Вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

В качестве ответа запишите количество, которое вернул запрос.

In [ ]:
query = f'''
            SELECT COUNT(*)
            FROM public.vacancies
            WHERE (LOWER(name) LIKE '%data scientist%'
                   OR LOWER(name) LIKE '%data science%'
                   OR LOWER(name) LIKE '%исследователь данных%'
                   OR (LOWER(name) LIKE '%ml%' AND LOWER(name) NOT LIKE '%html%')
                   OR LOWER(name) LIKE '%machine learning%'
                   OR LOWER(name) LIKE '%машинн%обучен%')
              AND LOWER(key_skills) LIKE '%python%';
         '''
df = pd.read_sql_query(query, connection)
df

/tmp/ipython-input-2066916185.py:12: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0    357

## Вывод
Python востребован больше чем SQL

## Задание 5
Сколько ключевых навыков в среднем указывают в вакансиях для DS?

Ответ округлите до двух знаков после точки-разделителя.

In [ ]:
query = f'''
           SELECT ROUND(AVG(skill_count), 2) AS avg_skills
FROM (
    SELECT
        CASE
            WHEN key_skills IS NULL OR key_skills = '' THEN 0
            ELSE ARRAY_LENGTH(STRING_TO_ARRAY(key_skills, ','), 1)
        END AS skill_count
    FROM public.vacancies
    WHERE (LOWER(name) LIKE '%data scientist%'
           OR LOWER(name) LIKE '%data science%'
           OR LOWER(name) LIKE '%исследователь данных%'
           OR (LOWER(name) LIKE '%ml%' AND LOWER(name) NOT LIKE '%html%')
           OR LOWER(name) LIKE '%machine learning%'
           OR LOWER(name) LIKE '%машинн%обучен%')
      AND LOWER(key_skills) LIKE '%python%'
) AS skills_subquery;
         '''
df = pd.read_sql_query(query, connection)
df


/tmp/ipython-input-230893277.py:19: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


avg_skills
0        1.03

## Задание 6
Напишите запрос, позволяющий вычислить, какую зарплату для DS в среднем указывают для каждого типа требуемого опыта (уникальное значение из поля experience).

При решении задачи примите во внимание следующее:

Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
Если заполнены оба поля с зарплатой, считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, его и считаем зарплатой по вакансии.
Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null).
Чтобы избежать этой ситуации, мы воспользуемся функцией coalesce, которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос select 1 + coalesce(null, 0).
Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа.

In [ ]:
query = f'''
            -- Общий запрос для средних зарплат по всем типам опыта
SELECT
    experience,
    ROUND(AVG(calculated_salary)) AS avg_salary
FROM (
    SELECT
        experience,
        CASE
            WHEN salary_from IS NOT NULL AND salary_to IS NOT NULL
                THEN (salary_from + salary_to) / 2.0
            ELSE COALESCE(salary_from, salary_to)
        END AS calculated_salary
    FROM public.vacancies
    WHERE (LOWER(name) LIKE '%data scientist%'
           OR LOWER(name) LIKE '%data science%'
           OR LOWER(name) LIKE '%исследователь данных%'
           OR (LOWER(name) LIKE '%ml%' AND LOWER(name) NOT LIKE '%html%')
           OR LOWER(name) LIKE '%machine learning%'
           OR LOWER(name) LIKE '%машинн%обучен%')
      AND (salary_from IS NOT NULL OR salary_to IS NOT NULL)
) AS salary_data
GROUP BY experience
ORDER BY experience;
         '''
df = pd.read_sql_query(query, connection)
df

/tmp/ipython-input-2171904813.py:26: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


experience  avg_salary
0         Более 6 лет    157933.0
1           Нет опыта     74643.0
2  От 1 года до 3 лет    144233.0
3       От 3 до 6 лет    256454.0

## Вывод
Прирост зарплат происходит с увеличением опыта работы. Однако у специалистов с опытом больше 6 лет зарплата снижается. Возможно в базе присутствует какая то аномалия

## Список из 10 городов где наиболе востребованы вакансии связанные с обработкой данных


In [ ]:
query = f'''
            SELECT
                a.name AS city,
                COUNT(v.id) AS data_vacancy_count
            FROM public.vacancies v
            JOIN public.areas a ON v.area_id = a.id
            WHERE
                LOWER(v.name) LIKE '%data scientist%'
                OR LOWER(v.name) LIKE '%data science%'
                OR LOWER(v.name) LIKE '%исследователь данных%'
                OR (LOWER(v.name) LIKE '%ml%' AND LOWER(v.name) NOT LIKE '%html%')
                OR LOWER(v.name) LIKE '%machine learning%'
                OR LOWER(v.name) LIKE '%машинн%обучен%'
            GROUP BY a.name
            ORDER BY data_vacancy_count DESC
            LIMIT 10;
         '''
df = pd.read_sql_query(query, connection)
df

/tmp/ipython-input-2400565467.py:18: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


city  data_vacancy_count
0           Москва                 225
1  Санкт-Петербург                  69
2      Новосибирск                  23
3           Алматы                  22
4  Нижний Новгород                  20
5           Казань                  19
6            Томск                  11
7   Ростов-на-Дону                  11
8            Минск                  11
9            Пермь                   8

## Вывод
Москва и Санкт-Петербург в списке лидеров по количеству вакансий связанных с обработкой данных.Это может говорить о том что основная  часть компаний сосредоточена именно там.

### Финал

In [ ]:
# не забываем закрыть соединение после окончания работы
connection.close()